README
  This modification of PlibC is made primary for microhttpd.
  It includes UTF-8 support, many fixes and enhancements.

	plibc - POSIX compliant libc extension for Windows



  Copyright (C) 2005 - 2012 Nils Durner
  Copyright (C) 2013 Evgeny Grin

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Comments are welcome.

	- Nils Durner <ndurner@users.sourceforge.net>


ChangeLog
2013-03-15  Evgeny Grin
  * Path functions respect buffer size

2013-03-14  Evgeny Grin
  * UTF-8 fixes
  * Real x64 compatibility - a lot of fixes
  * Fix missing declarations in public and private headers
  * random now use rand()

2012-05-05  Richard Boulton
  * pthreads compatibility

2012-04-20  LRN
  * UTF-8 Support

2011-11-30  LRN
  * msync()
  * MinGW64 fixes

2011-11-25  Nils Durner
  * nl_item
  * fix cross-platform build
  * improved random number generator
  * gmtime_r

2010-11-06	Nils Durner
	* mkfifo()
	* use MinGW printf

2010-06-22	Nils Durner
	* sysconf()

2010-01-17  Nils Durner
	* sockets functions now take/return ints instead of SOCKETs

2010-01-09	Nils Durner
	* strcasestr()

2010-01-07	Michael Lenaghan
	* shortcut.cc is now plain C

2009-12-01	Nils Durner <ndurner@users.sourceforge.net>
	* translate top level paths that do not end with "/"
	* define error_t
	
2009-11-19	Nils Durner <ndurner@users.sourceforge.net>
	* stpcpy()

2009-11-05	Nils Durner	<ndurner@users.sourceforge.net>
	* handle "~" at the beginning of a path
	* releasing 0.1.5

2009-09-24	Nils Durner <ndurner@users.sourceforge.net>
	* added SRANDOM()

2009-05-18	Nils Durner <ndurner@users.sourceforge.net>
	* avoid linking against libuuid
	* rename CTIME() to PLIBC_CTIME()
	* return errno style error code from getsockopt()
	* new: GetErrnoFromWinsockError()

2009-03-30	Nils Durner <ndurner@users.sourceforge.net>
	* renamed ctime(), ctime_r(), gettimeofday() and search.h functions to
	  _win_*()

2009-03-25	Nils Durner <ndurner@users.sourceforge.net>
	* added search functions (tsearch, lsearch, hsearch)

2009-01-29	Nils Durner <ndurner@users.sourceforge.net>
	* releasing 0.1.4

2008-08-17	Nils Durner <ndurner@users.sourceforge.net>
	* Vista compatibility
	* gethostbyname2()
	* various resolver functions
	* kill()
	* fix mkstemp()

2008-06-15	Nils Durner <ndurner@users.sourceforge.net>
	* plibc.ini can now override the parameters passed to plibc_init()

2006-01-12	Nils Durner	<ndurner@users.sourceforge.net>
	* plibc_conv_to_win_path(_ex)() no longer crashes if passed null pointers

2006-01-05  Nils Durner  <ndurner@users.sourceforge.net>
	* fstat() didn't work with pipes
	* support for "$HOME" in path names

2006-05-13	Nils Durner  <ndurner@users.sourceforge.net>
	* releasing 0.1.3

2006-05-08	Nils Durner  <ndurner@users.sourceforge.net>
  * strncasecmp(), strcasecmp

2006-04-18	Nils Durner  <ndurner@users.sourceforge.net>
	* stat64, lstat64

2006-12-03	Nils Durner  <ndurner@users.sourceforge.net>

	* support for locales in strptime
	* define MAP_FAILED
	* define EILSEQ and EOVERFLOW
	* strndup

2005-12-20  Nils Durner  <ndurner@users.sourceforge.net>

	* dereferencing symlinks/shortcuts always set errno = EINVAL
	* readlink() returned incorrect data sometimes

2005-09-04  Nils Durner  <ndurner@users.sourceforge.net>

	* Added check for ftruncate()
	* Fix symlink(): when overwriting an existing link, a link by the same name
	                 as the orginal file was created
	* Make header file compatible with gcc 2.95 (Unix)
	* Fix symlinks under Win9x
	* If newfile named a symlink, rename() didn't operate on the link itself
	* Symlink names were not expanded (dir/file -> dir\file.lnk)
	* Add support for "Open file" common dialog
	* fix mkstemp()
	* fix memory leak in symlink dereferencing

2005-07-08  Nils Durner  <ndurner@users.sourceforge.net>

	* releasing 0.1.0

2005-04-27  Nils Durner  <ndurner@users.sourceforge.net>

	* initial version.



src/path.c
/*
     This file is part of PlibC.
     (C) 2005, 2006, 2007, 2008 Nils Durner (and other contributing authors)

	   This library is free software; you can redistribute it and/or
	   modify it under the terms of the GNU Lesser General Public
	   License as published by the Free Software Foundation; either
	   version 2.1 of the License, or (at your option) any later version.
	
	   This library is distributed in the hope that it will be useful,
	   but WITHOUT ANY WARRANTY; without even the implied warranty of
	   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
	   Lesser General Public License for more details.
	
	   You should have received a copy of the GNU Lesser General Public
	   License along with this library; if not, write to the Free Software
	   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/**
 * @file src/path.c
 * @brief Translation Unix pathnames -> Windows pathnames
 */

#include "plibc_private.h"

extern char szuRootDir[_MAX_PATH + 1];
extern long luRootDirLen;
extern char szuHomeDir[_MAX_PATH + 2];
extern long luHomeDirLen;
extern char szuDataDir[_MAX_PATH + 1];
extern long luDataDirLen;
extern char szuUser[261];
extern char *_pszuOrg;
extern char *_pszuApp;

extern wchar_t szRootDir[_MAX_PATH + 1];
extern long lRootDirLen;
extern wchar_t szHomeDir[_MAX_PATH + 2];
extern long lHomeDirLen;
extern wchar_t szDataDir[_MAX_PATH + 1];
extern long lDataDirLen;
extern wchar_t szUser[261];
extern wchar_t *_pwszOrg;
extern wchar_t *_pwszApp;

/**
 * @brief Determine the Windows path of our / directory
 * @internal
 * @return Error code from winerror.h, ERROR_SUCCESS on success
 */
long _plibc_DetermineRootDir(void)
{
  wchar_t wszModule[_MAX_PATH], wszDrv[_MAX_DRIVE], wszDir[_MAX_DIR];
  wchar_t *pwszBin;
  long lDirLen;

  /* Get the path of the calling module.
     It should be located in one of the "bin" directories */
  GetModuleFileNameW(NULL, wszModule, MAX_PATH);
  _wsplitpath(wszModule, wszDrv, wszDir, NULL, NULL);

  lDirLen = wcslen(wszDir);
  pwszBin = wszDir + lDirLen - 5;
  if (lDirLen < 5 || wcsnicmp(pwszBin, L"\\bin\\", 5) != 0)
  {
    wchar_t wszRegPath[251];

    /* Get the installation path from the registry */
    lDirLen = _MAX_PATH - 1;
    _win_snwprintf(wszRegPath, 250, L"Software\\%s\\%s", _pwszOrg, _pwszApp);
    wszRegPath[250] = 0;

    if(QueryRegistryW
       (HKEY_CURRENT_USER, wszRegPath, L"InstallDir",
        szRootDir, &lDirLen) != ERROR_SUCCESS)
    {
      lDirLen = _MAX_PATH - 1;

      if(QueryRegistryW
         (HKEY_LOCAL_MACHINE, wszRegPath, L"InstallDir",
          szRootDir, &lDirLen) != ERROR_SUCCESS)
      {
        if (GetCurrentDirectoryW(sizeof(szRootDir), szRootDir) == 0)
          return ERROR_BAD_ENVIRONMENT;
        lDirLen = wcslen(szRootDir);
      }
    }
    wcscat(szRootDir, L"\\");
    lRootDirLen = lDirLen + 1;
    wszDrv[0] = 0;
  }
  else
  {
  	pwszBin[1] = 0;
  	lDirLen -= 4;
  }

  if(wszDrv[0])
  {
    wcscpy(szRootDir, wszDrv);
    lRootDirLen = 3 + lDirLen - 1;      /* 3 = strlen(szDir) */
    if(lRootDirLen > _MAX_PATH)
      return ERROR_BUFFER_OVERFLOW;

    wcscat(szRootDir, wszDir);
  }

  if (_plibc_utf8_mode == 1)
    wchartostr_buf (szRootDir, szuRootDir, _MAX_PATH + 1, CP_UTF8);
  else
    wchartostr_buf (szRootDir, szuRootDir, _MAX_PATH + 1, CP_ACP);
  luRootDirLen = strlen (szuRootDir);
  return ERROR_SUCCESS;
}

/**
 * @brief Determine the user's home directory
 * @internal
 * @return Error code from winerror.h, ERROR_SUCCESS on success
*/
long _plibc_DetermineHomeDir(void)
{
  wchar_t *lpwszProfile = _wgetenv(L"USERPROFILE");
  if(lpwszProfile != NULL && lpwszProfile[0] != 0)        /* Windows NT */
  {
    lHomeDirLen = wcslen(lpwszProfile);
    if(lHomeDirLen + 1 > _MAX_PATH)
      return ERROR_BUFFER_OVERFLOW;

    wcscpy(szHomeDir, lpwszProfile);
    if(szHomeDir[lHomeDirLen - 1] != L'\\')
    {
      szHomeDir[lHomeDirLen] = L'\\';
      szHomeDir[++lHomeDirLen] = 0;
    }
  }
  else
  {
    /* C:\My Documents */
    long lRet;

    lHomeDirLen = _MAX_PATH;
    lRet = QueryRegistryW(HKEY_CURRENT_USER,
                         L"Software\\Microsoft\\Windows\\CurrentVersion\\"
                         L"Explorer\\Shell Folders",
                         L"Personal", szHomeDir, &lHomeDirLen);

    if(lRet == ERROR_BUFFER_OVERFLOW)
      return ERROR_BUFFER_OVERFLOW;
    else if(lRet == ERROR_SUCCESS)
    {
      /* lHomeDirLen includes \0 */
      if (lHomeDirLen <= _MAX_PATH)
        wcscat(szHomeDir, L"\\");
      else
        return ERROR_BUFFER_OVERFLOW;
    }
    else
    {
      /* C:\Program Files\GNUnet\home\... */
      /* 5 = strlen("home\\") */
      lHomeDirLen = wcslen(szRootDir) + wcslen(szUser) + 5 + 1;

      if(_MAX_PATH < lHomeDirLen)
        return ERROR_BUFFER_OVERFLOW;

      wcscpy(szHomeDir, szRootDir);
      wcscat(szHomeDir, L"home\\");
      wcscat(szHomeDir, szUser);
      wcscat(szHomeDir, L"\\");
    }
  }

  if (_plibc_utf8_mode == 1)
    wchartostr_buf (szHomeDir, szuHomeDir, _MAX_PATH + 1, CP_UTF8);
  else
    wchartostr_buf (szHomeDir, szuHomeDir, _MAX_PATH + 1, CP_ACP);
  luHomeDirLen = strlen (szuHomeDir);
  return ERROR_SUCCESS;
}

long _plibc_DetermineProgramDataDir(void)
{
  long lRet;
  
  lDataDirLen = _MAX_PATH;
  lRet = QueryRegistryW(HKEY_LOCAL_MACHINE,
                       L"Software\\Microsoft\\Windows\\CurrentVersion\\"
                       L"Explorer\\Shell Folders",
                       L"Common AppData", szDataDir, &lDataDirLen);

  lDataDirLen += wcslen(_pwszApp) + 1 + wcslen(_pwszOrg) + 1; 
  if (lRet == ERROR_BUFFER_OVERFLOW || lDataDirLen > _MAX_PATH)
  {
    return ERROR_BUFFER_OVERFLOW;
  }
  else if (lRet == ERROR_SUCCESS)
  {
    wcscat(szDataDir, L"\\");
    wcscat(szDataDir, _pwszOrg);
    wcscat(szDataDir, L"\\");
    wcscat(szDataDir, _pwszApp);
    wcscat(szDataDir, L"\\");
  }
  else
    wcscpy(szDataDir, szRootDir);
  
  if (_plibc_utf8_mode == 1)
    wchartostr_buf (szDataDir, szuDataDir, _MAX_PATH + 1, CP_UTF8);
  else
    wchartostr_buf (szDataDir, szuDataDir, _MAX_PATH + 1, CP_ACP);
  luDataDirLen = strlen (szuDataDir);
  return ERROR_SUCCESS;
}

/**
 * @brief Convert a POSIX-sytle path to a Windows-style path
 * @param pszUnix POSIX path
 * @param pszWindows Windows path
 * @return Error code from winerror.h, ERROR_SUCCESS on success
*/
int plibc_conv_to_win_path(const char *pszUnix, char *pszWindows, size_t pszWindows_buff_length)
{
  return plibc_conv_to_win_path_ex(pszUnix, pszWindows, pszWindows_buff_length, 1);
}

int plibc_conv_to_win_pathw(const wchar_t *pszUnix, wchar_t *pszWindows, size_t pszWindows_buff_length)
{
  return plibc_conv_to_win_pathw_ex(pszUnix, pszWindows, pszWindows_buff_length, 1);
}

int plibc_conv_to_win_pathwconv(const char *pszUnix, wchar_t *pszWindows, size_t pszWindows_buff_length)
{
  wchar_t *pwszUnix;
  int r;
  r = strtowchar (pszUnix, &pwszUnix, CP_UTF8);
  if (r < 0)
    return r;
  r = plibc_conv_to_win_pathw_ex(pwszUnix, pszWindows, pszWindows_buff_length, 1);
  free (pwszUnix);
  return r;
}

int plibc_conv_to_win_pathwconv_ex(const char *pszUnix, wchar_t *pszWindows, size_t pszWindows_buff_length, int derefLinks)
{
  wchar_t *pwszUnix;
  int r;
  r = strtowchar (pszUnix, &pwszUnix, CP_UTF8);
  if (r < 0)
    return r;
  r = plibc_conv_to_win_pathw_ex(pwszUnix, pszWindows, pszWindows_buff_length, derefLinks);
  free (pwszUnix);
  return r;
}

/**
 * @brief Convert a POSIX-sytle path to a Windows-style path
 * @param pszUnix POSIX path
 * @param pszWindows Windows path
 * @param derefLinks 1 to dereference links
 * @return Error code from winerror.h, ERROR_SUCCESS on success
*/
int plibc_conv_to_win_pathw_ex(const wchar_t *pszUnix, wchar_t *pszWindows, size_t pszWindows_buff_length, int derefLinks)
{
  wchar_t *pSrc, *pDest;
  long iSpaceUsed;
  int iUnixLen;

  if (!pszUnix || !pszWindows)
    return ERROR_INVALID_PARAMETER;
  if (pszWindows_buff_length <= 0)
    return ERROR_BUFFER_OVERFLOW;

  iUnixLen = wcslen(pszUnix);

  /* Check if we already have a windows path */
  if((wcschr(pszUnix, L'\\') != NULL) || (wcschr(pszUnix, L':') != NULL))
  {
    if(iUnixLen > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    wcsncpy(pszWindows, pszUnix, pszWindows_buff_length);
  }

  /* Temp. dir? */
  if(wcsncmp(pszUnix, L"/tmp", 4) == 0)
  {
    iSpaceUsed = GetTempPathW(pszWindows_buff_length, pszWindows);
	if (iSpaceUsed == 0 )
      return ERROR_INVALID_PARAMETER;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + iSpaceUsed;
    pSrc = (wchar_t *) pszUnix + 4;
  }
  /* Bit bucket? */
  else if (wcsncmp(pszUnix, L"/dev/null", 9) == 0)
  {
    wcsncpy(pszWindows, L"nul", pszWindows_buff_length);
    iSpaceUsed = 3;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + lHomeDirLen;
    pSrc = (wchar_t *) pszUnix + 9;
  }
  /* Data directories */
  else if (wcsncmp(pszUnix, L"/etc", 4) == 0 ||
    wcsncmp(pszUnix, L"/com", 4) == 0 ||
    wcsncmp(pszUnix, L"/var", 4) == 0)
  {
    wcsncpy(pszWindows, szDataDir, pszWindows_buff_length);
    iSpaceUsed = lDataDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + lDataDirLen;
    pSrc = (wchar_t *) pszUnix + 1;
  }
  /* Is the unix path a full path? */
  else if(pszUnix[0] == L'/')
  {
    wcsncpy(pszWindows, szRootDir, pszWindows_buff_length);
    iSpaceUsed = lRootDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + lRootDirLen;
    pSrc = (wchar_t *) pszUnix + 1;
  }
  /* Home dir? */
  else if (pszUnix[0] == L'~')
  {
    wcsncpy(pszWindows, szHomeDir, pszWindows_buff_length);
    iSpaceUsed = lHomeDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + lHomeDirLen;
    pSrc = (wchar_t *) pszUnix + 1;
  }
  /* Home dir (env var)? */
  else if (wcsncmp(pszUnix, L"$HOME", 5) == 0)
  {
    wcsncpy(pszWindows, szHomeDir, pszWindows_buff_length);
    iSpaceUsed = lHomeDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + lHomeDirLen;
    pSrc = (wchar_t *) pszUnix + 5;  	
  }
  else
  {
    pDest = pszWindows;
    iSpaceUsed = 0;
    pSrc = (wchar_t *) pszUnix;
  }

  iSpaceUsed += wcslen(pSrc);
  if(iSpaceUsed + 1 > pszWindows_buff_length)
    return ERROR_BUFFER_OVERFLOW;

  /* substitute all slashes */
  while(*pSrc)
  {
    if(*pSrc == L'/')
      *pDest = L'\\';
    else
      *pDest = *pSrc;

    pDest++;
    pSrc++;
  }
  *pDest = 0;

  if (derefLinks)
    __win_derefw(pszWindows);
  else
  {
    /* The filename possibly refers to a symlink, but the .lnk extension may be
       missing.
        1. Check if the requested file seems to be a normal file
        2. Check if the file exists
         2.1. Yes: Finished
         2.2. No: Check if "filename.lnk" exists
          2.2.1 Yes: Append ".lnk" */
    if (wcsnicmp(pDest - 4, L".lnk", 4) != 0)
    {
      HANDLE h = CreateFileW(pszWindows, 0, FILE_SHARE_READ | FILE_SHARE_WRITE,
        NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
      if (h == INVALID_HANDLE_VALUE)
      {
        /* File doesn't exist, try shortcut */
        wchar_t *pLnk;
        int mal;
        
        if (iSpaceUsed + 5 > pszWindows_buff_length)
        {
          pLnk = malloc((iSpaceUsed + 5) * sizeof (wchar_t));
          wcscpy(pLnk, pszWindows);
          mal = 1;
        }
        else
        {
          pLnk = pszWindows;
          mal = 0;
        }
        wcscat(pLnk, L".lnk");
        
        h = CreateFileW(pLnk, 0, FILE_SHARE_READ | FILE_SHARE_WRITE,
          NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
          
        if (h != INVALID_HANDLE_VALUE)
        {
          /* Shortcut exists */
          CloseHandle(h);
          if (mal)
		  {
            /* Need to copy */
            if (iSpaceUsed + 5 <= pszWindows_buff_length)
              wcscpy(pszWindows, pLnk);
            else
            {
              free(pLnk);
              return ERROR_BUFFER_OVERFLOW;
            }
		  }
        }
        else
          pLnk[iSpaceUsed] = 0;   
        
        if (mal)
          free(pLnk);
      }
      else
        CloseHandle(h);
    }
  }

#if DEBUG_WINPROC
	{
		char szInfo[1001];

  	_win_snprintf(szInfo, 1000, "Posix path %S resolved to %S\n", pszUnix,
  		pszWindows);
  	szInfo[1000] = 0;
  	__plibc_panic(INT_MAX, szInfo);
	}
#endif

  return ERROR_SUCCESS;
}

int plibc_conv_to_win_path_ex(const char *pszUnix, char *pszWindows, size_t pszWindows_buff_length, int derefLinks)
{
  char *pSrc, *pDest;
  long iSpaceUsed;
  int iUnixLen;

  if (!pszUnix || !pszWindows)
    return ERROR_INVALID_PARAMETER;
  if (pszWindows_buff_length <= 0)
    return ERROR_BUFFER_OVERFLOW;

  iUnixLen = strlen(pszUnix);

  /* Check if we already have a windows path */
  if((strchr(pszUnix, '\\') != NULL) || (strchr(pszUnix, ':') != NULL))
  {
    if(iUnixLen > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    strncpy(pszWindows, pszUnix, pszWindows_buff_length);
  }

  /* Temp. dir? */
  if(strncmp(pszUnix, "/tmp", 4) == 0)
  {
    iSpaceUsed = GetTempPath(pszWindows_buff_length, pszWindows);
	if (iSpaceUsed == 0)
      return ERROR_INVALID_PARAMETER;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + iSpaceUsed;
    pSrc = (char *) pszUnix + 4;
  }
  /* Bit bucket? */
  else if (strncmp(pszUnix, "/dev/null", 9) == 0)
  {
    strncpy(pszWindows, "nul", pszWindows_buff_length);
    iSpaceUsed = 3;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + luHomeDirLen;
    pSrc = (char *) pszUnix + 9;
  }
  /* Data directories */
  else if (strncmp(pszUnix, "/etc", 4) == 0 ||
    strncmp(pszUnix, "/com", 4) == 0 ||
    strncmp(pszUnix, "/var", 4) == 0)
  {
    strncpy(pszWindows, szuDataDir, pszWindows_buff_length);
    iSpaceUsed = luDataDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + luDataDirLen;
    pSrc = (char *) pszUnix + 1;
  }
  /* Is the unix path a full path? */
  else if(pszUnix[0] == '/')
  {
    strncpy(pszWindows, szuRootDir, pszWindows_buff_length);
    iSpaceUsed = luRootDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + luRootDirLen;
    pSrc = (char *) pszUnix + 1;
  }
  /* Home dir? */
  else if (pszUnix[0] == '~')
  {
    strncpy(pszWindows, szuHomeDir, pszWindows_buff_length);
    iSpaceUsed = luHomeDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + luHomeDirLen;
    pSrc = (char *) pszUnix + 1;
  }
  /* Home dir (env var)? */
  else if (strncmp(pszUnix, "$HOME", 5) == 0)
  {
    strncpy(pszWindows, szuHomeDir, pszWindows_buff_length);
    iSpaceUsed = luHomeDirLen;
    if (iSpaceUsed > pszWindows_buff_length)
      return ERROR_BUFFER_OVERFLOW;
    pDest = pszWindows + luHomeDirLen;
    pSrc = (char *) pszUnix + 5;  	
  }
  else
  {
    pDest = pszWindows;
    iSpaceUsed = 0;
    pSrc = (char *) pszUnix;
  }

  iSpaceUsed += strlen(pSrc);
  if(iSpaceUsed + 1 > pszWindows_buff_length)
    return ERROR_BUFFER_OVERFLOW;

  /* substitute all slashes */
  while(*pSrc)
  {
    if(*pSrc == '/')
      *pDest = '\\';
    else
      *pDest = *pSrc;

    pDest++;
    pSrc++;
  }
  *pDest = 0;

  if (derefLinks)
    __win_deref(pszWindows);
  else
  {
    /* The filename possibly refers to a symlink, but the .lnk extension may be
       missing.
        1. Check if the requested file seems to be a normal file
        2. Check if the file exists
         2.1. Yes: Finished
         2.2. No: Check if "filename.lnk" exists
          2.2.1 Yes: Append ".lnk" */
    if (strnicmp(pDest - 4, ".lnk", 4) != 0)
    {
      HANDLE h = CreateFile(pszWindows, 0, FILE_SHARE_READ | FILE_SHARE_WRITE,
        NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
      if (h == INVALID_HANDLE_VALUE)
      {
        /* File doesn't exist, try shortcut */
        char *pLnk;
        int mal;
        
        if (iSpaceUsed + 5 > pszWindows_buff_length)
        {
          pLnk = malloc(iSpaceUsed + 5);
          strcpy(pLnk, pszWindows);
          mal = 1;
        }
        else
        {
          pLnk = pszWindows;
          mal = 0;
        }
        strcat(pLnk, ".lnk");
        
        h = CreateFile(pLnk, 0, FILE_SHARE_READ | FILE_SHARE_WRITE,
          NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
          
        if (h != INVALID_HANDLE_VALUE)
        {
          /* Shortcut exists */
          CloseHandle(h);
          if (mal)
          {
            /* Need to copy */
            if (iSpaceUsed + 5 <= pszWindows_buff_length)
              strcpy(pszWindows, pLnk);
            else
            {
              free(pLnk);
              return ERROR_BUFFER_OVERFLOW;
            }
          }
        }
        else
          pLnk[iSpaceUsed] = 0;   
        
        if (mal)
          free(pLnk);
      }
      else
        CloseHandle(h);
    }
  }

#if DEBUG_WINPROC
	{
		char szInfo[1001];

  	_win_snprintf(szInfo, 1000, "Posix path %s resolved to %s\n", pszUnix,
  		pszWindows);
  	szInfo[1000] = 0;
  	__plibc_panic(INT_MAX, szInfo);
	}
#endif

  return ERROR_SUCCESS;
}

/* end of path.c */
